pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go

```diff
@@ -0,0 +1,727 @@
+	conts := []string{}
+	pods := []string{}
+	for podID, contIDs := range pool.PodIDs {
+		podName := podID
+		if pod, ok := p.cch.LookupPod(podID); ok {
+			podName = pod.GetName()
+		}
+		pods = append(pods, fmt.Sprintf("%s (mCPU: %d, max=%d)", podName, p.getPodMilliCPU(podID), p.podMaxMilliCPU[podID]))
+		for _, contID := range contIDs {
+			if cont, ok := p.cch.LookupContainer(contID); ok {
+				conts = append(conts, cont.PrettyName())
+			} else {
+				conts = append(conts, podName+":"+contID)
+			}
+		}
+	}
+	s := fmt.Sprintf("Pool{Def.Name: %q, Instance: %d, CPUs: %s, Mems: %s, Def.MaxPods: %d, pods: %v, containers:%v}",
+		pool.Def.Name, pool.Instance, pool.CPUs, pool.Mems, pool.Def.MaxPods, pods, conts)
+	return s
+}
+
+// freePool removes an empty pod from a pool
+func (p *podpools) freePool(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	delete(pool.PodIDs, podID)
+	delete(p.podMaxMilliCPU, podID)
+}
+
+// trackPodCPU keeps track on pod's CPU requests.
+func (p *podpools) trackPodCPU(pod cache.Pod, pool *Pool) {
+	podID := pod.GetID()
+	current := p.getPodMilliCPU(podID)
+	if max, ok := p.podMaxMilliCPU[podID]; ok {
+		if max < current {
+			p.podMaxMilliCPU[podID] = current
+		}
+	} else {
+		p.podMaxMilliCPU[podID] = current
+	}
+	// Check overbooking
+	if cpuAvail := p.availableMilliCPUs(pool); cpuAvail < 0 {
+		log.Error("overbooked pool %q, cpuset:%s: %dm / %dm CPUs used, %d mCPU available", pool.PrettyName(), pool.CPUs, pool.CPUs.Size()*1000-int(cpuAvail), pool.CPUs.Size()*1000, cpuAvail)
+	}
+}
+
+// validatePodCPU compares max CPU requests against pool CPU capacity per pod.
+func (p *podpools) validatePodCPU(pod cache.Pod, pool *Pool) {
+	// Log pod configuration error if a pool has fixed amount of
+	// CPUs per pod but the pod failed to request the correct
+	// amount.
+	podID := pod.GetID()
+	if podmCPU, ok := p.podMaxMilliCPU[podID]; ok {
+		if pool.Def.MaxPods > 0 {
+			poolmCPUperPod := int64(pool.CPUs.Size() * 1000 / pool.Def.MaxPods)
+			mCPUerr := podmCPU - poolmCPUperPod
+			if mCPUerr < -podMilliCPUErrorMargin || mCPUerr > podMilliCPUErrorMargin {
+				podName := ""
+				if pod, ok := p.cch.LookupPod(podID); ok {
+					podName = pod.GetName()
+				}
+				log.Error("bad CPU requests: pod %q requested %d mCPUs, but in pool %q pods must request %d mCPUs.", podName, podmCPU, pool.Def.Name, poolmCPUperPod)
+			}
+		}
+	}
+}
+
+// getPodMilliCPU returns mCPUs requested by podID.
+func (p *podpools) getPodMilliCPU(podID string) int64 {
+	cpuRequested := int64(0)
+	for _, c := range p.cch.GetContainers() {
+		if c.GetPodID() == podID {
+			if reqCpu, ok := c.GetResourceRequirements().Requests[corev1.ResourceCPU]; ok {
+				cpuRequested += reqCpu.MilliValue()
+			}
+		}
+	}
+	return cpuRequested
+}
+
+// configNotify applies new configuration.
+func (p *podpools) configNotify(event pkgcfg.Event, source pkgcfg.Source) error {
+	log.Info("configuration %s", event)
+	if err := p.setConfig(podpoolsOptions); err != nil {
```

Comment:
Stupid question: what happens to the already running pods? Are they re-allocated somewhere/somehow?

Reply:
@klihub, working on this...

Reply:
Reconfiguration implemented.